go/pkg/casng/batching.go

```diff
@@ -314,12 +318,113 @@ func (u *uploader) writeBytes(ctx context.Context, name string, r io.Reader, siz
-	panic("not yet implemented")
+	contextmd.Infof(ctx, log.Level(1), "[casng] upload; reqs=%d", len(reqs))
+	defer contextmd.Infof(ctx, log.Level(1), "[casng] upload.done; reqs=%d", len(reqs))
+
+	var stats Stats
+
+	if len(reqs) == 0 {
+		return nil, stats, nil
+	}
+
+	var undigested []UploadRequest
+	digested := make(map[digest.Digest]UploadRequest)
+	var digests []digest.Digest
+	for _, r := range reqs {
+		if r.Digest.Hash == "" {
+			undigested = append(undigested, r)
+			continue
+		}
+		digested[r.Digest] = r
+		digests = append(digests, r.Digest)
+	}
+	missing, err := u.MissingBlobs(ctx, digests)
+	if err != nil {
+		return nil, stats, err
+	}
+	contextmd.Infof(ctx, log.Level(1), "[casng] upload; missing=%d, undigested=%d", len(missing), len(undigested))
+
+	reqs = undigested
+	for _, d := range missing {
+		reqs = append(reqs, digested[d])
+		delete(digested, d)
+	}
+	for d := range digested {
+		stats.BytesRequested += d.Size
+		stats.LogicalBytesCached += d.Size
+		stats.CacheHitCount += 1
+		stats.DigestCount += 1
+	}
+	if len(reqs) == 0 {
+		contextmd.Infof(ctx, log.Level(1), "[casng] upload: nothing is missing")
+		return nil, stats, nil
+	}
+
+	contextmd.Infof(ctx, log.Level(1), "[casng] upload: uploading; reqs=%d", len(reqs))
+	ch := make(chan UploadRequest)
+	resCh := u.streamPipe(ctx, ch)
+
+	u.clientSenderWg.Add(1)
+	go func() {
+		defer close(ch) // let the streamer terminate.
+		defer u.clientSenderWg.Done()
+
+		contextmd.Infof(ctx, log.Level(1), "[casng] upload.sender.start")
+		defer contextmd.Infof(ctx, log.Level(1), "[casng] upload.sender.stop")
+
+		for _, r := range reqs {
+			select {
+			case ch <- r:
+			case <-u.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	var uploaded []digest.Digest
+	for r := range resCh {
+		if r.Err != nil {
+			err = errors.Join(r.Err, err)
+		}
+		stats.Add(r.Stats)
+		// For pure stat responses, i.e. r.Digest is zero, CacheMissCount will be equal to 0.
+		if r.Stats.CacheMissCount > 0 {
+			uploaded = append(uploaded, r.Digest)
+		}
+	}
+
+	return uploaded, stats, err
 }
 
 // DigestTree returns the digest of the merkle tree for root.
 func (u *BatchingUploader) DigestTree(ctx context.Context, root impath.Absolute, slo symlinkopts.Options, exclude walker.Filter) (digest.Digest, Stats, error) {
-	panic("not yet implemented")
+	ch := make(chan UploadRequest)
+	resCh := u.streamPipe(ctx, ch)
+
+	req := UploadRequest{Path: root, SymlinkOptions: slo, Exclude: exclude, ctx: ctx, digestOnly: true}
+	select {
+	// In both cases, the channel must be closed before proceeding beyong the select statement to ensure
+	// the proper closure of resCh.
+	case ch <- req:
+		close(ch)
+	case <-ctx.Done():
+		close(ch)
+		return digest.Digest{}, Stats{}, ctx.Err()
+	}
+
+	var stats Stats
+	var err error
+	for r := range resCh {
+		stats.Add(r.Stats)
+		err = errors.Join(r.Err, err)
+	}
+	rootNode := u.Node(req)
```

Comment:
`u.Node(req)` returns the interface `proto.Message` which will never be nil right? You probably need to do `!u.Node(req).ProtoReflect().IsValid()` instead of `rootNode == nil` or try to cast to each of the possible concrete types?

Reply:
Since the method is called from a sibling one on the same struct, I'm trusting the implementation to return either `nil` (not found) or a valid struct. Verbose guards seem an overkill considering this is technically internal (to the struct) code.

Reply:
That is reasonable. Maybe add to the docs on Node() that it will either return a valid struct or nil. That can be in one of the follow up changes though

Reply:
I thought I already did: https://github.com/bazelbuild/remote-apis-sdks/blob/master/go/pkg/casng/uploader.go#L224-L227
What can I add to improve it?